barrister/runtime.py

```diff
@@ -423,6 +423,243 @@ def _call(self, context):
+        """
+
+        if self.log.isEnabledFor(logging.DEBUG):
+            self.log.debug("Request: %s" % str(req))
+
+        if isinstance(req, list):
+            if len(req) < 1:
+                d = defer.fail(err_response(None, ERR_INVALID_REQ, "Invalid Request. Empty batch."))
+            else:
+                # run the batch call collecting the responses
+                d = defer.DeferredList([self._call_and_format(r, props) for r in req])
+        else:
+            d = self._call_and_format(req, props)
+
+        def log_response(response):
+            if self.log.isEnabledFor(logging.DEBUG):
+                self.log.debug("Response: %s" % str(response))
+            return response
+
+        d.addBoth(log_response)
+        return d
+
+    def _call_and_format(self, req, props=None):
+        """
+        Invokes a single request against a handler using _call() and traps any errors,
+        formatting them using _err().  If the request is successful it is wrapped in a
+        JSON-RPC 2.0 compliant dict with keys: 'jsonrpc', 'id', 'result'.
+
+        :Parameters:
+          req
+            A single dict representing a single JSON-RPC request
+          props
+            Application defined properties to set on RequestContext for use with filters.
+            For example: authentication headers.  Must be a dict.
+        """
+        if not isinstance(req, dict):
+            return err_response(None, ERR_INVALID_REQ,
+                                "Invalid Request. %s is not an object." % str(req))
+
+        reqid = None
+        if "id" in req:
+            reqid = req["id"]
+
+        if props is None:
+            props = {}
+        context = RequestContext(props, req)
+
+        if self.filters:
+            for f in self.filters:
+                f.pre(context)
+
+        if context.error:
+            return context.error
+
+        d = self._call(context)
+
+        def makeResponse(result):
+            return { "jsonrpc": "2.0", "id": reqid, "result": result }
+
+        def handleRpcException(failure):
+            failure.trap(RpcException)
+            e = failure.value
+            return err_response(reqid, e.code, e.msg, e.data)
+
+        d.addCallbacks(makeResponse, handleRpcException)
+
+        def handleUnexpectedExceptions(failure):
+            failure.trap(Exception)
+            self.log.exception("Error processing request: %s" % str(req))
+            return err_response(reqid, ERR_UNKNOWN,
+                                "Server error. Check logs for details.",
+                                data={'exception': str(failure.value)})
+
+        d.addErrback(handleUnexpectedExceptions)
+
+        if self.filters:
+            def postHook(hook, result):
+                context.response = result
+                hook(context)
+                return result
+
+            for f in self.filters:
+                d.addBoth(lambda r: postHook(f.post, r))
```

Comment:
Understood.
I would have written it slightly differently, to keep closer to the original: [untested]
```
if self.filters:
    def set_response(resp):
        context.response = resp
        return resp

    def apply_filter(resp, hook):
        hook(context)
        return resp

    d.addCallback(set_response)
    for f in self.filters:
        d.addCallback(apply_filter, f.post)
```
But that is mostly a matter of opinion. :)